### ISSUE TITLE
Support emitting typescript decorator metadata

### ISSUE THREAD

**Issue by SirJosh3917** [evanw/esbuild#257]
For projects such as TypeORM, it is necessary for decorator metadata to get emitted. For example, let's take TypeORM's [`typescript-example`](https://github.com/typeorm/typescript-example) project. Running the project normally yields the following output.
```shell
git clone https://github.com/typeorm/typescript-example --depth 1
cd typescript-example
yarn && yarn start
```
![image](https://user-images.githubusercontent.com/9854865/87383435-39306e00-c567-11ea-9d85-431b4e243c67.png)

Attempting to use ESBuild successfully compiles it with warnings, but fails with a runtime error.
```shell
esbuild --bundle --outfile=index.js --target=esnext --platform=node --format=cjs --strict .\src\index.ts
```
![image](https://user-images.githubusercontent.com/9854865/87383744-1783b680-c568-11ea-9a39-df61f39be630.png)

**Comment by evanw** [comment]
The `emitDecoratorMetadata` flag is intentionally not supported. It relies on running the TypeScript type checker, which relies on running the TypeScript compiler, which is really slow. Given that esbuild's primary purpose is speed, I will not be integrating the TypeScript compiler into esbuild. I will also not be rewriting the TypeScript type checker in Go since that would be a massive ongoing maintenance burden. It's possible that you could write an esbuild plugin for this when esbuild's plugin API materializes (see #111) but that would likely eliminate most of the speed advantage of using esbuild in the first place. You're probably better off using another tool instead of esbuild if you need to do this.

**Comment by SirJosh3917** [comment]
What do you think about adding some way to short circuit the type checking process, such as an `@esbuild-type [number/string/object/boolean/function]` annotation? That could be a solution for this situation specifically at least, but I'm not sure about the versatility of the idea at a larger scope. It's certainly not an ideal solution and is at best a hack, but if it could be useful in other areas that'd be a great side effect.

**Comment by evanw** [comment]
That annotation seems too much of a special-case to put into esbuild core to me. Besides, if you want to be able to do this:

```js
class Foo {
  @decorator
  // @esbuild-type number
  prop: number
}
```

couldn't you just do something like this instead?

```js
class Foo {
  @decorator
  @esbuildType(Number)
  prop: number
}
```

where `esbuildType` is just a function that forwards to the TypeScript metadata function:

```js
let esbuildType = t => __metadata("design:type", t)
```

I think that would do what you're trying to accomplish without needing to extend esbuild at all. In fact it's even shorter than the comment annotation form.

I know implementing just those types would be part of the feature, but I don't think it's a good idea to claim to support it and then work differently in subtle ways that silently break things. Even those primitive types can be propagated through arbitrary aliases and type expressions in the type system and can probably be imported from other files. Getting all of that correct means re-implementing the TypeScript type checker.

**Comment by SirJosh3917** [comment]
Alright, that's completely understandable. Thanks for taking the time to run through the idea :+1:

**Comment by axe-me** [comment]
I saw swc actually support `emitDecoratorMetadata`. This is how it did it: https://github.com/swc-project/swc/pull/939/files
Would esbuild so something similar? 
lots of backend packages are built on top of this flag. it's kinda annoying not able to use esbuild to improve backend DX. 

Thanks.

**Comment by evanw** [comment]
This is in scope for swc because they are [trying to replicate the whole TypeScript type system in swc](https://github.com/swc-project/swc/issues/571). Putting a type system into esbuild is firmly out of scope for esbuild. If you need this feature you should use swc/spack instead (or maybe use swc as a plugin for esbuild?).

**Comment by shadowtime2000** [comment]
@evanw I am pretty sure that the SWC project to replicate `tsc` is a closed source one, while the open source parts of SWC support `emitDecoratorMetadata` so that may not be the case, but I am not 100% sure.

**Comment by evanw** [comment]
Ah weird. Ok well I guess never mind. I missed that part. Looks like this is the relevant comment: https://github.com/swc-project/swc/issues/571#issuecomment-745746895

> But it’s not open sourced, and I’m not going to open source it.

**Comment by shadowtime2000** [comment]
@evanw Do you think it may still be possible considering how the parts of SWC which act just like Babel are able to do it?

**Comment by thomaschaaf** [comment]
I'm currently trying to write an esbuild plugin which will try to build with esbuild and compile only the files containing docorators with tsc again. This should result in faster build times as usually not all files use decorators. See https://github.com/evanw/esbuild/issues/915#issuecomment-791758649

**Comment by thomaschaaf** [comment]
I have released the following plugin: (esbuild-plugin-tsc)[https://github.com/thomaschaaf/esbuild-plugin-tsc] which allows you to now use `emitDecoratorMetadata` with esbuild.

**Comment by vikasgarghb** [comment]
@evanw Is there any workaround to this issue?
@SirJosh3917 What solution did you go for?

**Comment by evanw** [comment]
> @evanw Is there any workaround to this issue?

One workaround is in the comment immediately preceding yours: you can write an esbuild plugin that runs the TypeScript transpiler on source code files containing decorators before passing the files to esbuild.

**Comment by SirJosh3917** [comment]
> @SirJosh3917 What solution did you go for?
I opted to not use esbuild, since I needed it.

**Comment by robbie-c** [comment]
Just as an experiment I wanted to test how sophisticated the type inference is when creating this metadata. TLDR it doesn't actually do much resolution of complex types at all, and doing something good enough for most use cases could have ok performance.

Input file (an example TypeORM table)
```typescript
// snipped imports

@Entity("some_table")
class SomeTable extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column()
  num: number;

  @Column()
  something_hard: Pick<{ a: string; b: number }, "a">["a"]; // string

  @OneToOne(() => DBDeviceUUID)
  lazy_relation: Promise<DBDeviceUUID>;

  @OneToOne(() => DBDeviceUUID)
  eager_relation: DBDeviceUUID;

  @Column()
  a_set: Set<string>;

  @Column()
  a_map: Map<string, string>;

  @Column()
  a_bigint: BigInteger;
}

const x: SomeTable["something_hard"] =
  "just proving that this is just a string";
```

output file
```javascript
// snipped defintions of metadata and decorate, and snipped requires

let SomeTable = class SomeTable extends typeorm_1.BaseEntity {
};
__decorate([
    (0, typeorm_1.PrimaryGeneratedColumn)("uuid"),
    __metadata("design:type", String)
], SomeTable.prototype, "id", void 0);
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", Number)
], SomeTable.prototype, "num", void 0);
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", Object)
], SomeTable.prototype, "something_hard", void 0);
__decorate([
    (0, typeorm_1.OneToOne)(() => DeviceUUID_1.DBDeviceUUID),
    __metadata("design:type", Promise)
], SomeTable.prototype, "lazy_relation", void 0);
__decorate([
    (0, typeorm_1.OneToOne)(() => DeviceUUID_1.DBDeviceUUID),
    __metadata("design:type", DeviceUUID_1.DBDeviceUUID)
], SomeTable.prototype, "eager_relation", void 0);
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", Set)
], SomeTable.prototype, "a_set", void 0);
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", Map)
], SomeTable.prototype, "a_map", void 0);
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", Object)
], SomeTable.prototype, "a_bigint", void 0);
SomeTable = __decorate([
    (0, typeorm_1.Entity)("some_table")
], SomeTable);
const x = "just proving that this is just a string";
```

Just to put this in a neater format
| type | design:type |
| -- | -- |
| `string` | String |
| `number` | Number|
| `Pick<{ a: string; b: number }, "a">["a"]` (string) | Object |
| `Promise<DBDeviceUUID>` | Promise |
| `DBDeviceUUID` | DBDeviceUUID |
| `Set` | Set |
| `Map` | Map |
| `BigInteger` | Object |

I was actually surprised that it didn't get `something_hard`. My personal opinion is that if tsc can't do this, then any other implementation of reflect doesn't need to either. This would mean that instead of checking whether e.g. a type eventually resolves to string, it could just check whether it's literally the characters `string`.

**Comment by imcm7** [comment]
Better not to do it, than to do it that way. Shit

**Comment by segevfiner** [comment]
@evanw [swc](https://swc.rs/docs/configuration/compilation#jsctransformdecoratormetadata) supports this, and I doubt they run the full TypeScript compiler, it is likely the actual transform is simpler then you think and doesn't really require complex type resolution. [tsup](https://github.com/egoist/tsup) runs swc as a workaround for this not being supported in esbuild, but it would obviously be nicer if this was supported natively.

**Comment by evanw** [comment]
I tried looking into the current rules. It looks like they are [here](https://github.com/microsoft/TypeScript/blob/25a4f9e53595fe944ada7cc747ec393ce0bc2c6e/src/compiler/transformers/typeSerializer.ts#L253). These rules imply that:

1. Type syntax would need to be parsed (currently esbuild treats types as whitespace)
2. Type identifiers need to be resolved by a type checker (e.g. even `declare` types)
3. Post-resolution type attributes must be inspected (e.g. is the type numeric, does the interface have a call signature)

Here is an example:

```ts
// foo.ts
declare let dec: any
import type { ns } from './bar'
class Foo {
  @dec fn(e: ns.E): ns.Fn { throw 0 }
}
```

```ts
// bar.ts
export namespace ns {
  export declare enum E {}
  export interface Fn { x(): void }
  export interface Fn extends Base {}
  type Base = [1] | [false] extends Array<number> ? Y : Call
  interface Y { y(): void }
  interface Call {
    (): void // This makes the return type "Function"
  }
}
```

You're supposed to get this:

```js
class Foo {
    fn(e) { throw 0; }
}
__decorate([
    dec,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Number]),
    __metadata("design:returntype", Function)
], Foo.prototype, "fn", null);
```

The `paramtypes` include `Number` because `ns.E` is resolved to an enum type. This involves doing cross-file type resolution and can't be done at run-time since `E` is just a type declaration and isn't actually a value. For a more complex example, the `returntype` is `Function` instead of `Object` because `ns.Fn` (after being merged) extends from `Base`, and `Base` has type `Call` because while `[1]` extends `Array<number>`, `[false]` does not extend `Array<number>` so the union `[1] | [false]` does not extend `Array<number>`, and `Call` has a call signature.

Obviously it's possible write a type checker for all of this (`tsc` is a proof of that). But it still means writing a type checker for a Turing-complete type system. I agree that it would be nice if esbuild contained such a type checker. Then we wouldn't need `tsc` anymore! But then I'd have to write and maintain a TypeScript-compatible type checker, which is a huge undertaking and is not something I want to do (thus why I consider this to be out of scope).

Obviously it's also possible to write a simpler type checker that ignores the complexity of TypeScript and checks a different type language instead. That appears to be the approach swc has taken. However, I don't want esbuild to take that approach. That involves signing up for a constant stream of issues (as people would rightly expect esbuild's type system to behave like TypeScript instead of esbuildScript) and it wouldn't be fair to users who would have to deal with subtle and unexpected bugs.

**Comment by segevfiner** [comment]
Thanks for the detailed write up! Guess we can hope stuff will move away from using due to the newer stable decorators...

**Comment by segevfiner** [comment]
For now, guess I'll just use `tsup` with it's supported of using `swc` to transpile decorator metadata, including it's caveats of course...

**Comment by Aylur** [comment]
> I will also not be rewriting the TypeScript type checker in Go since that would be a massive ongoing maintenance burden.

@evanw Now that TypeScript is being ported to Go by Microsoft themselves, do you think there might be a chance to revisit this issue in the future?

**Comment by quinnjr** [comment]
Bringing this back up because the reason it was closed doesn't really hold anymore.

The close was that emitting design:paramtypes needs a real type checker, and you weren't going to rewrite the TypeScript type checker in Go and then own it forever. Fair on both counts. But Microsoft went and wrote one, it's Go, and it already does this exact transform.

Where it lives, pinned at [70420d4](https://github.com/microsoft/typescript-go/tree/70420d41286e0822894436d4454ef044d9658ec5):

- [internal/transformers/tstransforms/metadata.go](https://github.com/microsoft/typescript-go/blob/70420d41286e0822894436d4454ef044d9658ec5/internal/transformers/tstransforms/metadata.go), the transform itself, about 14KB
- [internal/transformers/tstransforms/typeserializer.go](https://github.com/microsoft/typescript-go/blob/70420d41286e0822894436d4454ef044d9658ec5/internal/transformers/tstransforms/typeserializer.go), design:type / design:paramtypes / design:returntype
- [internal/checker/emitresolver.go](https://github.com/microsoft/typescript-go/blob/70420d41286e0822894436d4454ef044d9658ec5/internal/checker/emitresolver.go), the checker-backed printer.EmitResolver
- [internal/compiler/emitter.go#L104-L134](https://github.com/microsoft/typescript-go/blob/70420d41286e0822894436d4454ef044d9658ec5/internal/compiler/emitter.go#L104-L134), where it goes into the transform chain, gated on options.EmitDecoratorMetadata

The transform doesn't actually want the checker, it wants the [printer.EmitResolver](https://github.com/microsoft/typescript-go/blob/70420d41286e0822894436d4454ef044d9658ec5/internal/printer/emitresolver.go) interface, mostly GetTypeReferenceSerializationKind and the symbol accessibility helpers. The checker is just what happens to implement it.

It's all under internal/, which sounds like a dealbreaker and isn't. tsgolint and ttsc both drive tsgo's checker through go:linkname shims, and tsgolint codegens them ([tools/gen_shims/main.go](https://github.com/typescript-eslint/tsgolint/blob/main/tools/gen_shims/main.go)) across ast, checker, compiler, core, scanner, tsoptions, tspath, vfs. So this is linkable today, pinned, without waiting on anybody. I've asked upstream about a cleaner path ([discussion #481](https://github.com/microsoft/typescript-go/discussions/481)) but it isn't a prerequisite.

Costs, to be upfront about them:

- Binary size and init. Linking the checker is not free, so it'd have to be opt-in and only paid when emitDecoratorMetadata is actually set.
- Apache-2.0 into MIT. Fine to depend on, but it means carrying a NOTICE.
- Shims are maintained against a pinned tsgo, so bumps are manual.

To be clear on scope: this is metadata emit behind the tsconfig flag, delegating to upstream's implementation. No type checking in esbuild, no type resolution for you to own, nothing to keep in sync as TS moves.

Would you take a PR for this? I'd rather ask than write one you'd turn down on principle.
